UI/CUIControls.cpp

```diff
@@ -1930,81 +1931,123 @@ void ProductionInfoPanel::DoLayout() {
+                                           const GG::Clr& bar_color,
+                                           const GG::Clr& bg_color,
                                            const GG::Clr& outline_color) :
     Control(GG::X0, GG::Y0, GG::X1, GG::Y1, GG::NO_WND_FLAGS),
-    m_total_turns(std::max(1, total_turns)),
-    m_turns_completed(std::max(0.0, std::min<double>(turns_completed, m_total_turns))),
-    m_total_cost(total_cost),
-    m_turn_spending(turn_spending),
-    m_bar_color(bar_color),
-    m_background(background),
-    m_outline_color(outline_color)
-{}
+    m_num_segments(num_segments),
+    m_perc_completed(percent_completed),
+    m_perc_predicted(percent_predicted),
+    m_clr_bar(bar_color),
+    m_clr_bg(bg_color),
+    m_clr_outline(outline_color)
+{
+    // validate percentage values
+    if (m_perc_completed < 0.0f || m_perc_predicted < 0.0f ||
+        (m_perc_completed + m_perc_predicted) > 1.0f)
+    {
+        WarnLogger() << "Values not within percent range, clamping: "
+                     << m_perc_completed << ", " << m_perc_predicted;
+        boost::algorithm::clamp(m_perc_completed, 0.0f, 1.0f);
+        boost::algorithm::clamp(m_perc_predicted, 0.0f, 1.0f);
+    }
+}
 
 void MultiTurnProgressBar::Render() {
-    GG::Pt ul = UpperLeft(), lr = LowerRight();
-    bool segmented = true;
-    if (Width() / m_total_turns < 3)
-        segmented = false;
-
-    // draw background over whole area
-    FlatRectangle(ul, lr, m_background, m_outline_color, 1);
-
-    // draw predicted bar of next turn
-    GG::X completed_bar_width(std::max(1.0, Value(Width() * m_turns_completed / m_total_turns)));
-    GG::X predicted_bar_width(Value(Width() * (m_turn_spending / m_total_cost)));
-
-    if (predicted_bar_width > 3) {
-        GG::Pt predicted_bar_ul(ul.x + completed_bar_width - 1, ul.y);
-        GG::Pt predicted_bar_lr;
-        if ((completed_bar_width + predicted_bar_width) > Width()) // cut if beyond right border
-            predicted_bar_lr = GG::Pt(ul.x + Width(), lr.y);
-        else
-            predicted_bar_lr = GG::Pt(ul.x + completed_bar_width + predicted_bar_width, lr.y);
-        FlatRectangle(predicted_bar_ul, predicted_bar_lr, GG::LightColor(m_bar_color), m_outline_color, 1);
+    GG::Pt ul(UpperLeft());
+    GG::Pt lr(LowerRight());
+    GG::Y bottom(lr.y);
+    GG::Rect border_rect({GG::X1, GG::Y1}, {GG::X1, GG::Y1});
+
+    // background
+    GG::GL2DVertexBuffer bg_verts;
+    BufferStoreRectangle(bg_verts, {ul, lr}, border_rect);
+
+    // define completed and predicted bar sizes
+    GG::X comp_width(Width() * m_perc_completed);
+    GG::X pred_width(Width() * m_perc_predicted);
+    // maximum lower right points to lower right of control
+    GG::Rect comp_rect(ul, {std::min(lr.x, ul.x + comp_width), bottom});
+    GG::Rect pred_rect({std::max(ul.x, comp_rect.lr.x - 1), ul.y},
+                       {std::min(lr.x, comp_rect.lr.x + pred_width), bottom});
+
+    // predicted progress
```

Comment:
this is better, but storing the vertices in member variables that don't need to be refilled each frame would be preferable. or does doing this result in the prerender issues you mentioned?